### PR TITLE
[Security] Preserve URI fragment in HttpUtils::generateUri()

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/23675
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by iltar** [comment]
This way of url generating causes more issues as it simply throws in all attributes: iltar/http-bundle#25

**Comment by chalasr** [comment]
@iltar we should open another issue for that

File: src/Symfony/Component/Security/Http/HttpUtils.php [hunk 1]
```diff
@@ -150,7 +150,16 @@ public function generateUri($request, $path)
         // fortunately, they all are, so we have to remove entire query string
         $position = strpos($url, '?');
         if (false !== $position) {
+            $fragment = '';
```
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Http/HttpUtils.php]: Could we use http://php.net/parse_url here instead of this custom logic?

File: src/Symfony/Component/Security/Http/HttpUtils.php [hunk 2]
```diff
@@ -150,7 +150,16 @@ public function generateUri($request, $path)
         // fortunately, they all are, so we have to remove entire query string
```
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Http/HttpUtils.php]: Should we update this comment too?

Comment (chalasr) [reply src/Symfony/Component/Security/Http/HttpUtils.php]: dedicated comment added just before fragment handling

File: src/Symfony/Component/Security/Http/HttpUtils.php (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Http/HttpUtils.php]: better thanks

File: src/Symfony/Component/Security/Http/HttpUtils.php [hunk 3]
```diff
@@ -150,7 +150,12 @@ public function generateUri($request, $path)
...
         $position = strpos($url, '?');
         if (false !== $position) {
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Http/HttpUtils.php]: Why is that necessary? What about a path like `/foo/bar#fragment`?

Comment (chalasr) [reply src/Symfony/Component/Security/Http/HttpUtils.php]: Fragment gets stripped only if there is a query string, otherwise the urlGenerator result is returned directly

Comment (xabbuh) [reply src/Symfony/Component/Security/Http/HttpUtils.php]: Thanks for the clarification, but maybe we should add a test for that too to prevent regressions?

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/Security/Http/HttpUtils.php (hunk 3 above)
Comment (iltar) [reply src/Symfony/Component/Security/Http/HttpUtils.php]: In most cases there is _always_ a query string, because it uses the attributes from the request to build it.

Comment (chalasr) [reply src/Symfony/Component/Security/Http/HttpUtils.php]: @xabbuh Test added

**Review by javiereguiluz — APPROVED** [review]
